pkg/controller/servicebindingrequest/binding.go

```diff
@@ -0,0 +1,322 @@
+	if err != nil {
+		return RequeueOnConflict(err)
+	}
+
+	// appending finalizer, should be later removed upon resource deletion
+	sbr.SetFinalizers(append(sbr.GetFinalizers(), Finalizer))
+	if _, err = b.updateServiceBindingRequest(sbr); err != nil {
+		return NoRequeue(err)
+	}
+
+	b.Logger.Info("All done!")
+	return Done()
+}
+
+// setApplicationObjects replaces the Status's equivalent field.
+func (b *ServiceBinder) setApplicationObjects(
+	sbrStatus *v1alpha1.ServiceBindingRequestStatus,
+	objs []*unstructured.Unstructured,
+) {
+	names := []string{}
+	for _, obj := range objs {
+		names = append(names, fmt.Sprintf("%s/%s", obj.GetNamespace(), obj.GetName()))
+	}
+	sbrStatus.ApplicationObjects = names
+}
+
+// buildPlan creates a new plan.
+func buildPlan(
+	ctx context.Context,
+	dynClient dynamic.Interface,
+	sbr *v1alpha1.ServiceBindingRequest,
+) (*Plan, error) {
+	planner := NewPlanner(ctx, dynClient, sbr)
+	return planner.Plan()
+}
+
+// InvalidOptionsErr is returned when ServiceBinderOptions are not valid.
+var InvalidOptionsErr = errors.New("invalid options")
+
+// BuildServiceBinder creates a new binding manager according to options.
+func BuildServiceBinder(options *ServiceBinderOptions) (*ServiceBinder, error) {
+	if !options.Valid() {
+		return nil, InvalidOptionsErr
+	}
+
+	// objs groups all extra objects related to the informed SBR
+	objs := make([]*unstructured.Unstructured, 0)
+
+	// plan is a source of information regarding the binding process
+	ctx := context.Background()
+	plan, err := buildPlan(ctx, options.DynClient, options.SBR)
+	if err != nil {
+		return nil, err
+	}
+
+	rs := plan.GetCRs()
+	// append all SBR related CRs
+	objs = append(objs, rs...)
+
+	// retriever is responsible for gathering data related to the given plan.
+	retriever := NewRetriever(options.DynClient, plan, options.EnvVarPrefix)
+
+	// read bindable data from the specified resources
+	if options.DetectBindingResources {
+		err := retriever.ReadBindableResourcesData(&plan.SBR, rs)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// read bindable data from the CRDDescription found by the planner
+	for _, r := range plan.GetRelatedResources() {
+		err = retriever.ReadCRDDescriptionData(r.CR, r.CRDDescription)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// gather retriever's read data
+	// TODO: do not return error
+	retrievedData, err := retriever.Get()
+	if err != nil {
+		return nil, err
+	}
+
+	// gather related secret, again only appending it if there's a value.
+	secret := NewSecret(options.DynClient, plan)
+	secretObj, found, err := secret.Get()
+	if err != nil {
+		return nil, err
+	}
+	if found {
+		objs = append(objs, secretObj)
+	}
+
+	return &ServiceBinder{
+		Logger:    options.Logger,
+		Binder:    NewBinder(ctx, options.Client, options.DynClient, options.SBR, retriever.volumeKeys),
+		DynClient: options.DynClient,
+		SBR:       options.SBR,
+		Objects:   objs,
+		Data:      retrievedData,
```

Comment:
Initially, we had a retrievedCache as well, that captures the hierarchical format of retrievedData. Are we capturing that or omitting that?

Reply:
The hierarchical data format might be useful only when collecting and aggregating the configuration values and/or secrets; `retrievedData` is the aggregated information that will be persisted in a later stage.

Reply:
I am referring to this issue https://github.com/redhat-developer/service-binding-operator/issues/272 which was later fixed in https://github.com/redhat-developer/service-binding-operator/pull/281 .

Reply:
So we are keeping that. The behavior introduced in #281 hasn't been changed in this PR, and `ServiceBinder` will use the same `Retriever` implementation being currently used.